src/ctap/client_pin.rs

```diff
@@ -167,34 +116,32 @@ pub enum PinPermission {
 }
 
 pub struct ClientPin {
-    key_agreement_key: crypto::ecdh::SecKey,
-    pin_uv_auth_token: [u8; PIN_TOKEN_LENGTH],
+    pin_protocol_v1: PinProtocol,
     consecutive_pin_mismatches: u8,
     permissions: u8,
     permissions_rp_id: Option<String>,
 }
 
 impl ClientPin {
     pub fn new(rng: &mut impl Rng256) -> ClientPin {
-        let key_agreement_key = crypto::ecdh::SecKey::gensk(rng);
-        let pin_uv_auth_token = rng.gen_uniform_u8x32();
         ClientPin {
-            key_agreement_key,
-            pin_uv_auth_token,
+            pin_protocol_v1: PinProtocol::new(rng),
             consecutive_pin_mismatches: 0,
             permissions: 0,
             permissions_rp_id: None,
         }
     }
 
+    /// Checks the given encrypted PIN hash against the stored PIN hash.
+    ///
     /// Decrypts the encrypted pin_hash and compares it to the stored pin_hash.
     /// Resets or decreases the PIN retries, depending on success or failure.
     /// Also, in case of failure, the key agreement key is randomly reset.
     fn verify_pin_hash_enc(
         &mut self,
         rng: &mut impl Rng256,
         persistent_store: &mut PersistentStore,
-        aes_dec_key: &crypto::aes256::DecryptionKey,
+        shared_secret: &mut dyn SharedSecret,
```

Comment:
why do we need a `mut SharedSecret`? Encryption and decryption shouldn't alter the original data, right?
This comment also applies to the methods above of course.